File: seeker/snippet/queue.py
#date: 2022-08-25T16:53:18Z
#url: https://api.github.com/gists/4540fb821d21adee5df29609aec131bb
#owner: https://api.github.com/users/jermspeaks

# A "first-in-last-out" data structure
class QueueEmptyError(IndexError):
    """Attempt to pop an empty stack."""

class Queue:
    def __init__(self):
        self._list = []

    def enqueue(item):
        # Add an item to the back of the queue
        self._list.insert(0,item)

    def dequeue():
        # Remove an item from the front of the queue
        if not self._list:
            raise StackEmptyError
        else:
            self._list.pop()

    def peek():
        # Return first item of the queue
        self

#   isEmpty() {
#     // Returns a boolean if the queue is empty
#   }

#   get length() {
#     // Return the length of the queue
#   }
# }

# const queue = new Queue();
# console.log(queue);
# console.log(queue.length);

File: seeker/snippet/queue.go
//date: 2023-11-02T16:33:56Z
//url: https://api.github.com/gists/be12e78061eddcf69b89d56587641a87
//owner: https://api.github.com/users/mmtdebojit

package iterator

import (
	"fmt"
)

type Queue struct {
	items []interface{}
}

func (q *Queue) Enqueue(item interface{}) {
	q.items = append(q.items, item)
}

func (q *Queue) Dequeue() interface{} {
	if len(q.items) == 0 {
		return nil
	}
	item := q.items[0]
	q.items = q.items[1:]
	return item
}

func (q *Queue) IsEmpty() bool {
	return len(q.items) == 0
}

func (q *Queue) Size() int {
	return len(q.items)
}

func main() {
	queue := Queue{}

	queue.Enqueue(1)
	queue.Enqueue(2)
	queue.Enqueue(3)

	fmt.Println("Queue size:", queue.Size())

	for !queue.IsEmpty() {
		item := queue.Dequeue()
		fmt.Println("Dequeued:", item)
	}

	fmt.Println("Queue size:", queue.Size())
}
